core/corehttp/readers.go

```diff
@@ -0,0 +1,134 @@
+package corehttp
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/ipfs/go-ipfs/commands/files"
+	"github.com/ipfs/go-ipfs/importer"
+	chunk "github.com/ipfs/go-ipfs/importer/chunk"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+// helper functions for the gateway handler
+
+func (i *gatewayHandler) newDagFromReader(r io.Reader) (*dag.Node, error) {
+	// TODO(cryptix): change and remove this helper once PR1136 is merged
+	// return ufs.AddFromReader(i.node, r.Body)
+	return importer.BuildDagFromReader(
+		r, i.node.DAG, chunk.DefaultSplitter, importer.BasicPinnerCB(i.node.Pinning.GetManual()))
+}
+
+// add file, or directory recursively
+func (i *gatewayHandler) addFileRecursive(file files.File) (*dag.Node, error) {
+	if file.IsDirectory() {
+		return i.addDirRecursive(file)
+	}
+	return i.newDagFromReader(file)
+}
+
+// add a directory recursively
+func (i *gatewayHandler) addDirRecursive(file files.File) (*dag.Node, error) {
```

Comment:
agreed this would be nice. I was hoping the addFile/addDir functions would have been exported from the `core/commands`, but they're not and are tied to that params struct, so it would have taken more reorganization. Also the functions in `core/commands` do more than we need for the gateway, like outputting progress bars, pinning, and ignoring hidden files (I've taken the opinion that if you send a tar ball, the whole thing will get added, hidden files indeed). 

So, do you think we should add params for dealing with hidden files and pinning to the gateway?

Part of my hesitation in doing this (and in trying to use the functions in `core/commands`) was issues like https://github.com/ipfs/go-ipfs/issues/1322 which seem to suggest there is much greater reorg coming to the gateway to support the api. I figured deduplication would come then. Thoughts?